datafusion/core/src/physical_plan/sorts/sort.rs

```diff
@@ -253,11 +261,23 @@ impl ExternalSorter {
     /// Appends an unsorted [`RecordBatch`] to `in_mem_batches`
     ///
     /// Updates memory usage metrics, and possibly triggers spilling to disk
-    async fn insert_batch(&mut self, input: RecordBatch) -> Result<()> {
+    async fn insert_batch(&mut self, mut input: RecordBatch) -> Result<()> {
         if input.num_rows() == 0 {
             return Ok(());
         }
 
+        let mut batch_sorted = false;
+        if self
+            .fetch
+            .map_or(false, |f| f <= input.num_rows() && f <= 100)
+        {
+            // Eagerly sort the batch to potentially reduce the number of rows
+            // after applying the fetch parameter.
+            // Currently only applied for fetch of 100 rows or less.
+            input = sort_batch(&input, &self.expr, self.fetch)?;
```

Comment:
I wonder whether you could recover part of the perf difference by concar + sorting only once every `n` (say every 10) batches. The selectivity of the limit and total work to be performed is much bigger for sorting 81920 vs 8192 rows, also the merging to be performed will be over fewer batches.

Reply:
Yeah, that's worth trying to benchmark as well. It could also be something like every `n` unsorted rows or every `size` unsorted bytes to accommodate for variability in batch row count.

Reply:
Yes, definitely.

Reply:
Fwiw, I did try out a variety of approaches along these lines, but still failed to eliminate the perf difference. There's probably a combination of parameters that is just right and could yield better results, but with the native Top-K operator in the works that likely won't be necessary.